rasa/core/featurizers/tracker_featurizers.py

```diff
@@ -528,100 +606,172 @@ def slice_state_history(
+        example_labels = []
+        example_entities = []
 
         logger.debug(
-            "Creating states and action examples from "
+            "Creating states and {} label examples from "
             "collected trackers (by {}({}))..."
-            "".format(type(self).__name__, type(self.state_featurizer).__name__)
+            "".format(
+                self.LABEL_NAME,
+                type(self).__name__,
+                type(self.state_featurizer).__name__,
+            )
         )
         pbar = tqdm(
             trackers,
             desc="Processed trackers",
             disable=rasa.shared.utils.io.is_logging_disabled(),
         )
         for tracker in pbar:
-            states = self._create_states(
-                tracker, domain, omit_unset_slots=omit_unset_slots
-            )
 
-            states_length_for_action = 0
-            entity_data = {}
-            for event in tracker.applied_events():
-                if isinstance(event, UserUttered):
-                    entity_data = self._entity_data(event)
+            for states, label, entities in self._example_iterator(
+                tracker, domain, omit_unset_slots=omit_unset_slots
+            ):
 
-                if not isinstance(event, ActionExecuted):
+                if self._check_example_cache(tracker, states, label):
                     continue
 
-                states_length_for_action += 1
+                example_states.append(states)
+                example_labels.append(label)
+                example_entities.append(entities)
+
+                pbar.set_postfix({f"# {self.LABEL_NAME}": f"{len(example_labels):d}"})
+
+        self._cleanup_example_iterator()
+        self._remove_user_text_if_intent(example_states)
+
+        logger.debug(f"Created {len(example_states)} {self.LABEL_NAME} examples.")
+
+        return example_states, example_labels, example_entities
+
+    def _setup_example_iterator(self) -> None:
+        """Create set for filtering out duplicated training examples."""
+        if self.remove_duplicates:
+            self.hashed_examples = set()
+
+    def _example_iterator(
+        self,
+        tracker: DialogueStateTracker,
+        domain: Domain,
+        omit_unset_slots: bool = False,
+    ) -> Iterator[Tuple[List[State], List[Text], List[Dict[Text, Any]]]]:
+        """Create an iterator over the training examples that will be created
+        from the provided tracker.
+
+        Returns:
+            An iterator over state, labels, entity tag tuples
+        """
+
+        tracker_states = self._create_states(
+            tracker, domain, omit_unset_slots=omit_unset_slots
+        )
+
+        label_index = 0
+        entity_data = {}
+        for event in tracker.applied_events():
+            if isinstance(event, UserUttered):
+                entity_data = self._entity_data(event)
+
+            elif isinstance(event, ActionExecuted):
+
+                label_index += 1
```

Comment:
I guess this doesn't go directly to you, but I gotta say that it's not trivial to understand / verify that this relationship between `tracker_states` and the events works out in that way that you only advance the quasi pointer to the position in `tracker_states` for every `ActionExecuted` event and not for other types

Reply:
Agree with both of these comments. I found this part pretty confusing. @dakshvar22 if this isn't already in the doc, would this be a good candidate for the rasa 3.0 todos?